File: src/aico/commands/drop.py
import sys
from pathlib import Path

from aico.exceptions import InvalidInputError
from aico.fs import validate_input_paths
from aico.session_loader import load_active_session


def drop(
    file_paths: list[Path],
) -> None:
    session = load_active_session()

    current_files = set(session.data.context_files)
    valid_rels, outside_root_errors = validate_input_paths(session.root, file_paths, require_file_exists=False)
    errors_found = outside_root_errors

    files_to_drop: list[str] = []
    for rel in valid_rels:
        if rel in current_files:
            files_to_drop.append(rel)
            print(f"Dropped file from context: {rel}")
        else:
            print(f"Error: File not in context: {rel}", file=sys.stderr)
            errors_found = True

    if files_to_drop:
        new_context = sorted(current_files - set(files_to_drop))
        session.persistence.update_view_metadata(context_files=new_context)

    if errors_found:
        raise InvalidInputError("One or more files described could not be dropped.")

File: src/commands/drop.rs
use crate::exceptions::AicoError;
use crate::fs::validate_input_paths;
use crate::session::Session;
use std::path::PathBuf;

pub fn run(file_paths: Vec<PathBuf>) -> Result<(), AicoError> {
    let mut session = Session::load_active()?;

    // 1. Validate (require_exists = false)
    // Note: Python drop checks validity but allows files missing from disk to be dropped from context
    let (valid_rels, has_errors) = validate_input_paths(&session.root, &file_paths, false);

    // 2. Modify View
    let mut changed = false;

    // We track explicit errors for files not in context
    let mut context_errors = false;

    for rel in valid_rels {
        if let Some(pos) = session.view.context_files.iter().position(|x| x == &rel) {
            session.view.context_files.remove(pos);
            println!("Dropped file from context: {}", rel);
            changed = true;
        } else {
            eprintln!("Error: File not in context: {}", rel);
            context_errors = true;
        }
    }

    // 3. Save
    if changed {
        session.save_view()?;
    }

    if has_errors || context_errors {
        return Err(AicoError::InvalidInput(
            "One or more files could not be dropped.".into(),
        ));
    }

    Ok(())
}
